### PR TITLE
parity with eth2-merge-kurtosis-module

### PR DESCRIPTION
**Pull request by h4ck3rk3y**

### CODE REVIEW

**Comment by h4ck3rk3y** [comment]
adjust to the new smothened add_service

**Comment by h4ck3rk3y** [comment]
use the import module syntax

**Comment by h4ck3rk3y** [comment]
Will fix readme in a separate PR

**Comment by h4ck3rk3y** [comment]
Created #15 #14

**Comment by h4ck3rk3y** [comment]
finalization does happen but we can't do assertions on it yet

image

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+		el_client_context,
+		mev_boost_context,
+		log_level,
+		extra_beacon_params,
+	)
+
+	beacon_service = add_service(beacon_node_service_id, beacon_config)
+
+	# TODO check whether its 200, 206 or 503 like golang
+	define_fact(service_id = beacon_node_service_id, fact_name = BEACON_HEALTH_FACT_NAME, fact_recipe = struct(method= "GET", endpoint = "/eth/v1/node/health", content_type = "application/json", port_id = BEACON_HTTP_PORT_ID))
```
Comment (gbouv) [inline src/participant_network/cl/lighthouse/lighthouse_launcher.star:102]: We should maybe create our own named struct for those object. `fact_recipe(method = "GET", ...)` would look a bit more ironed IMO

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+	beacon_metrics_url = "{0}:{1}".format(beacon_service.ip_address, beacon_metrics_port.number)
+
+	validator_metrics_port = validator_service.ports[VALIDATOR_METRICS_PORT_ID]
+	validator_metrics_url = "{0}:{1}".format(validator_service.ip_address, validator_metrics_port.number)
+
+	beacon_node_metrics_info = new_cl_node_metrics_info(beacon_node_service_id, METRICS_PATH, beacon_metrics_url)
+	validator_node_metrics_info = new_cl_node_metrics_info(validator_node_service_id, METRICS_PATH, validator_metrics_url)
+	nodes_metrics_info = [beacon_node_metrics_info, validator_node_metrics_info]
+
+	result = new_cl_client_context(
```
Comment (gbouv) [inline src/participant_network/cl/lighthouse/lighthouse_launcher.star:136]: nit: could do
```
return new_cl_client_context(
        ...
```

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+			RUST_BACKTRACE_ENVVAR_NAME: RUST_FULL_BACKTRACE_KEYWORD
+		},
+	)
+
+
+
+def new_lighthouse_launcher(cl_genesis_data):
+	return struct(
+		genesis_data = cl_genesis_data,
+	)
```
Comment (gbouv) [inline src/participant_network/cl/lighthouse/lighthouse_launcher.star:300]: newline

File: src/participant_network/cl/lodestar/lodestar_launcher.star [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+PRIVATE_IP_ADDRESS_PLACEHOLDER = "KURTOSIS_IP_ADDR_PLACEHOLDER"
+
+BEACON_ENR_FACT_NAME = "beacon-enr-fact"
+BEACON_HEALTH_FACT_NAME = "beacon-health-fact"
+
+USED_PORTS = {
+	TCP_DISCOVERY_PORT_ID: new_port_spec(DISCOVERY_PORT_NUM, TCP_PROTOCOL),
+	UDP_DISCOVERY_PORT_ID: new_port_spec(DISCOVERY_PORT_NUM, UDP_PROTOCOL),
+	HTTP_PORT_ID:         new_port_spec(HTTP_PORT_NUM, TCP_PROTOCOL),
```
Comment (gbouv) [inline src/participant_network/cl/lodestar/lodestar_launcher.star:39]: Alignment is weird here

File: src/participant_network/cl/lodestar/lodestar_launcher.star [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+			node_keystore_files.files_artifact_uuid: VALIDATOR_KEYS_MOUNT_DIRPATH_ON_SERVICE_CONTAINER
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
+	)
+
+
+def new_lodestar_launcher(cl_genesis_data):
+	return struct(
+		cl_genesis_data = cl_genesis_data,
+	)
```
Comment (gbouv) [inline src/participant_network/cl/lodestar/lodestar_launcher.star:264]: newline

File: src/participant_network/cl/lodestar/lodestar_launcher.star [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = cmd_args,
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH_ON_SERVICE_CONTAINER,
+			node_keystore_files.files_artifact_uuid: VALIDATOR_KEYS_MOUNT_DIRPATH_ON_SERVICE_CONTAINER
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/cl/lodestar/lodestar_launcher.star:257]: `s/privaite/private`

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+		image = image,
+		ports = BEACON_USED_PORTS,
+		cmd_args = cmd_args,
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNTPOINT_ON_CLIENTS
+		},
+		env_vars = {
+			RUST_BACKTRACE_ENVVAR_NAME: RUST_FULL_BACKTRACE_KEYWORD
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/cl/lighthouse/lighthouse_launcher.star:231]: `s/privaite/private`

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+		image = image,
+		ports = VALIDATOR_USED_PORTS,
+		cmd_args = cmd_args,
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNTPOINT_ON_CLIENTS,
+			node_keystore_files.files_artifact_uuid: VALIDATOR_KEYS_MOUNTPOINT_ON_CLIENTS,
+		},
+		env_vars = {
+			RUST_BACKTRACE_ENVVAR_NAME: RUST_FULL_BACKTRACE_KEYWORD
+		},
```
Comment (gbouv) [inline src/participant_network/cl/lighthouse/lighthouse_launcher.star:292]: why don't we set the private ip placeholder here?

File: src/participant_network/cl/nimbus/nimbus_launcher.star [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+VALIDATOR_KEYS_MOUNTPOINT_ON_CLIENT = "/validator-keys"
+
+# Port IDs
+TCP_DISCOVERY_PORT_ID = "tcp-discovery"
+UDP_DISCOVERY_PORT_ID = "udp-discovery"
+HTTP_PORT_ID         = "http"
+METRICS_PORT_ID      = "metrics"
+
+# Port nums
+DISCOVERY_PORT_NUM = 9000
```
Comment (gbouv) [inline src/participant_network/cl/nimbus/nimbus_launcher.star:19]: Same here and above, alignment is weird

File: src/participant_network/cl/nimbus/nimbus_launcher.star [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = [cmd_str],
+		entry_point_args = ["sh", "-c"],
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNTPOINT_ON_CLIENT,
+			node_keystore_files.files_artifact_uuid: VALIDATOR_KEYS_MOUNTPOINT_ON_CLIENT
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/cl/nimbus/nimbus_launcher.star:212]: `s/privaite/private`

File: src/participant_network/cl/prysm/prysm_launcher.star [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+		cmd_args.extend([param for param in extra_params])
+
+	return struct(
+		image = beacon_image,
+		ports = BEACON_NODE_USED_PORTS,
+		cmd_args = cmd_args,
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH_ON_SERVICE_CONTAINER,
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/cl/prysm/prysm_launcher.star:222]: `s/privaite/private`

File: src/participant_network/cl/prysm/prysm_launcher.star [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+		cmd_args.append(("--http-mev-relay{0}".format(mev_boost_endpoint(mev_boost_context))))
+
+	if len(extra_params) > 0:
+		# we do the for loop as otherwise its a proto repeated array
+		cmd_args.extend([param for param in extra_params])
+
+	return struct(
+		image = beacon_image,
+		ports = BEACON_NODE_USED_PORTS,
+		cmd_args = cmd_args,
```
Comment (gbouv) [inline src/participant_network/cl/prysm/prysm_launcher.star:218]: it's a bit weird that for prism we set `entry_point_args = ["sh", "-c"],` but here we don't isn't it?

File: src/participant_network/cl/prysm/prysm_launcher.star [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+	return struct(
+		image = validator_image,
+		ports = VALIDATOR_NODE_USED_PORTS,
+		cmd_args = cmd_args,
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH_ON_SERVICE_CONTAINER,
+			node_keystore_files.files_artifact_uuid:             VALIDATOR_KEYS_MOUNT_DIRPATH_ON_SERVICE_CONTAINER,
+			prysm_password_artifact_uuid:          PRYSM_PASSWORD_MOUNT_DIRPATH_ON_SERVICE_CONTAINER,			
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/cl/prysm/prysm_launcher.star:282]: `s/privaite/private`

File: src/participant_network/cl/teku/teku_launcher.star [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = [cmd_str],
+		entry_point_args = ["sh", "-c"],
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH_ON_SERVICE_CONTAINER,
+			node_keystore_files.files_artifact_uuid: VALIDATOR_KEYS_DIRPATH_ON_SERVICE_CONTAINER
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/cl/teku/teku_launcher.star:214]: `s/privaite/private`

File: src/participant_network/el/besu/besu_launcher.star [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = [launch_node_command_str],
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_DIRPATH_ON_CLIENT_CONTAINER
+		},
+		entry_point_args = ENTRYPOINT_ARGS,
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/el/besu/besu_launcher.star:131]: `s/privaite/private`

File: src/participant_network/el/besu/besu_launcher.star [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+USED_PORTS = {
+	RPC_PORT_ID: new_port_spec(RPC_PORT_NUM, TCP_PROTOCOL),
+	WS_PORT_ID: new_port_spec(WS_PORT_NUM, TCP_PROTOCOL),
+	TCP_DISCOVERY_PORT_ID: new_port_spec(DISCOVERY_PORT_NUM, TCP_PROTOCOL),
+	UDP_DISCOVERY_PORT_ID: new_port_spec(DISCOVERY_PORT_NUM, UDP_PROTOCOL),
+	ENGINE_HTTP_RPC_PORT_ID: new_port_spec(ENGINE_HTTP_RPC_PORT_NUM, TCP_PROTOCOL),
+	ENGINE_WS_RPC_PORT_ID: new_port_spec(ENGINE_WS_RPC_PORT_NUM, TCP_PROTOCOL)
+}
+
+ENTRYPOINT_ARGS = ["sh", "-c"]
```
Comment (gbouv) [inline src/participant_network/el/besu/besu_launcher.star:37]: I guess we could do the same in all the other laucnhers

File: src/participant_network/el/erigon/erigon_launcher.star [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+		entry_point_args = ENTRYPOINT_ARGS,
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
+	)
+
+
+def new_erigon_launcher(network_id, el_genesis_data):
+	return struct(
+		network_id = network_id,
+		el_genesis_data = el_genesis_data,
+	)
```
Comment (gbouv) [inline src/participant_network/el/erigon/erigon_launcher.star:143]: newline

File: src/participant_network/el/erigon/erigon_launcher.star [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = [command_arg_str],
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH
+		},
+		entry_point_args = ENTRYPOINT_ARGS,
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/el/erigon/erigon_launcher.star:135]: `s/privaite/private`

File: src/participant_network/el/geth/geth_launcher.star [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = [command_str],
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH,
+			prefunded_geth_keys_artifact_uuid: PREFUNDED_KEYS_MOUNT_DIRPATH
+		},
+		entry_point_args = ENTRYPOINT_ARGS,
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER
```
Comment (gbouv) [inline src/participant_network/el/geth/geth_launcher.star:183]: `s/privaite/private`

File: src/participant_network/el/geth/geth_launcher.star [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+	)
+
+
+def new_geth_launcher(network_id, el_genesis_data, prefunded_geth_keys_artifact_uuid, prefunded_account_info):
+	return struct(
+		network_id = network_id,
+		el_genesis_data = el_genesis_data,
+		prefunded_account_info = prefunded_account_info,
+		prefunded_geth_keys_artifact_uuid = prefunded_geth_keys_artifact_uuid,
+	)
```
Comment (gbouv) [inline src/participant_network/el/geth/geth_launcher.star:193]: newline

File: src/participant_network/el/nethermind/nethermind_launcher.star [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+		command_args.extend([param for param in extra_params])
+
+	return struct(
+		image = image,
+		ports = USED_PORTS,
+		cmd_args = command_args,
+		files = {
+			genesis_data.files_artifact_uuid: GENESIS_DATA_MOUNT_DIRPATH
+		},
+		privaite_ip_address_placeholder = PRIVATE_IP_ADDRESS_PLACEHOLDER,
```
Comment (gbouv) [inline src/participant_network/el/nethermind/nethermind_launcher.star:125]: `s/privaite/private` and same about the entrypoints args

File: src/participant_network/participant_network.star [hunk 22]
```diff
@@ -52,12 +124,72 @@ def launch_participant_network(num_participants, network_params):
...
+			mev_boost_service_id = MEV_BOOST_SERVICE_ID_PREFIX.format(1)
+			mev_boost_context = launch_mevboost(mev_boost_launcher, mev_boost_service_id, network_params.network_id)
+
+		all_mevboost_contexts.append(mev_boost_context)
+
+		cl_client_context = None
+
+		if index == 0:
+			cl_client_context = launch_method(cl_launcher, cl_service_id, participant.cl_client_image, participant.cl_client_log_level, global_log_level, CL_CLIENT_CONTEXT_BOOTNODE,
+              el_client_context, mev_boost_context, new_cl_node_validator_keystores, participant.beacon_extra_params, participant.validator_extra_params)
```
Comment (gbouv) [inline src/participant_network/participant_network.star:169]: Not sure about the formatting here, but maybe putting one arg per line would make this a bit more readable

File: src/participant_network/participant_network.star [hunk 23]
```diff
@@ -52,12 +124,72 @@ def launch_participant_network(num_participants, network_params):
...
+		else:
+			boot_cl_client_ctx = all_cl_client_contexts[0]
+			cl_client_context = launch_method(cl_launcher, cl_service_id, participant.cl_client_image, participant.cl_client_log_level, global_log_level, boot_cl_client_ctx,
+              el_client_context, mev_boost_context, new_cl_node_validator_keystores, participant.beacon_extra_params, participant.validator_extra_params)
```
Comment (gbouv) [inline src/participant_network/participant_network.star:173]: and same here

**Review by gbouv — APPROVED** [review]
A few nits but the code looks great 👏  Trusting you that it works fine

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star (hunk 1 above)
Comment (h4ck3rk3y) [reply src/participant_network/cl/lighthouse/lighthouse_launcher.star:102]: @victorcolombo  is reworking how this works underneath 
https://www.notion.so/kurtosistech/One-off-HTTP-Request-cf290802339b4753b4bdde0946e1cd31

File: src/participant_network/cl/lighthouse/lighthouse_launcher.star (hunk 8 above)
Comment (h4ck3rk3y) [reply src/participant_network/cl/lighthouse/lighthouse_launcher.star:291]: none of the cmd args, env vars or entrypoint args use the place holder, so it needs no replacement. It picks the default which is fine.

File: src/participant_network/cl/prysm/prysm_launcher.star (hunk 13 above)
Comment (h4ck3rk3y) [reply src/participant_network/cl/prysm/prysm_launcher.star]: btw all of these continued to work as my PRIVATE_IP_ADDRESS_PLACEHOLDER was the default. We should fail loudly for struct keys that shouldn't be there. I thi

File: src/participant_network/cl/prysm/prysm_launcher.star (hunk 12 above)
Comment (h4ck3rk3y) [reply src/participant_network/cl/prysm/prysm_launcher.star]: I am just following - https://github.com/kurtosis-tech/eth2-merge-kurtosis-module/blob/fda6eec31c2c28db4630080960a6d59378997015/kurtosis-module/impl/participant_network/cl/prysm/prysm_launcher.go

seems like prysm and nimbus are broken on the main repo anyway. I have issues for those
https://github.com/kurtosis-tech/eth2-merge-kurtosis-module/issues/156

File: src/participant_network/el/besu/besu_launcher.star (hunk 16 above)
Comment (h4ck3rk3y) [reply src/participant_network/el/besu/besu_launcher.star:37]: i think in some of them the containers own entrypoint isn't enough, we override it and then launch via cmd. for others the entrypoint is enough. I'll make sure that I use a constant for the entrypoint where ever its overriden in the original module.

**Pull request merged**